Compléments pour PR 19307 : mise à jour de l'ensemble de la page (#2)
v-gb/translated-content: files/fr/mozilla/add-ons/webextensions/internationalization/index.md

---
title: Internationalisation
slug: Mozilla/Add-ons/WebExtensions/Internationalization
l10n:
  sourceCommit: 16d1903feb387e054a502c20f35ccb12d823dedb
---

{{AddonSidebar}}

L'API [WebExtensions](/fr/docs/Mozilla/Add-ons/WebExtensions) dispose d'un module pour [l'internationalisation](/fr/docs/Glossary/Internationalization) des extensions&nbsp;: [`i18n`](/fr/docs/Mozilla/Add-ons/WebExtensions/API/i18n). Dans cet article, nous allons explorer ses fonctionnalités et illustrer son utilisation par un exemple pratique. L'API `i18n` pour les extensions web est similaire aux bibliothèques JavaScript tierces d'internationalisation.

> > **Note :** L'exemple d'extension présenté dans cet article, [`notify-link-clicks-i18n`](https://github.com/mdn/webextensions-examples/tree/main/notify-link-clicks-i18n), est disponible sur GitHub. Suivez le code source pendant que vous parcourez les sections suivantes.

Reviewer: ```suggestion
> **Note :** L'exemple d'extension présenté dans cet article, [`notify-link-clicks-i18n`](https://github.com/mdn/webextensions-examples/tree/main/notify-link-clicks-i18n), est disponible sur GitHub. Suivez le code source pendant que vous parcourez cette page.
```

## Anatomie d'une extension internationalisée

> Une extension internationalisée peut contenir les mêmes fonctionnalités que n'importe quelle autre extension ([scripts d'arrière-plan](/fr/docs/Mozilla/Add-ons/WebExtensions/Anatomy_of_a_WebExtension#scripts_darrière-plan), [scripts de contenu](/fr/docs/Mozilla/Add-ons/WebExtensions/Anatomy_of_a_WebExtension#scripts_de_contenu), etc.), mais elle possède du contenu supplémentaire lui permettant de basculer entre différentes locales. Les éléments ajoutés pour l'internationalisation sont listés dans l'arborescence suivante&nbsp;:

Reviewer: ```suggestion
Une extension internationalisée peut contenir les mêmes fonctionnalités que n'importe quelle autre extension ([scripts d'arrière-plan](/fr/docs/Mozilla/Add-ons/WebExtensions/Anatomy_of_a_WebExtension#scripts_darrière-plan), [scripts de contenu](/fr/docs/Mozilla/Add-ons/WebExtensions/Anatomy_of_a_WebExtension#scripts_de_contenu), etc.), mais avec du contenu supplémentaire lui permettant de basculer entre différentes locales. Les éléments ajoutés pour l'internationalisation sont résumés dans l'arborescence suivante&nbsp;:
```

- `répertoire-racine-de-l'extension/`
  - `_locales`
    - `en`
      - `messages.json`&nbsp;: un fichier qui contiendra les chaînes de caractères en anglais
    - `fr_FR`
      - `messages.json`&nbsp;: un fichier qui contiendra les chaînes de caractères en français
    - `de`
      - `messages.json`&nbsp;: un fichier qui contiendra les chaînes de caractères en allemand
    - etc.
  - `manifest.json`&nbsp;: des métadonnées relatives aux locales sont ajoutées au [manifeste](/fr/docs/Mozilla/Add-ons/WebExtensions/manifest.json)

>   - `fichierJavaScript.js`&nbsp;: dans les scripts, on ajoutera généralement une logique pour récupérer la locale du navigateur, certains messages spécifiques pour chaque locale, etc.

Reviewer: ```suggestion
  - `fichierJavaScript.js`&nbsp;: de la logique pour récupérer la locale du navigateur, récupérer les messages spécifiques à la locale courante, etc.
```

>   - `mesStyles.css`&nbsp;: le code CSS pourra utiliser des paramètres pour chaque locale si nécessaire.

Reviewer: ```suggestion
  - `mesStyles.css`&nbsp;: les règles CSS peuvent changer de comportement suivant la locale
```

Examinons chacune de ces fonctionnalités supplémentaires. Chacune des sections qui suit représente une étape à suivre pour l'internationalisation de votre extension.

## Fournir des chaînes localisées dans `_locales`

> L'internationalisation nécessite de fournir des chaînes traduites pour les différentes locales qu'on souhaite prendre en charge. Pour les extensions, le répertoire `_locales`, présent à la racine de l'extension, contient un répertoire pour chaque locale, nommé grâce à [l'étiquette de langue correspondante](https://fr.wikipedia.org/wiki/%C3%89tiquette_d%27identification_de_langues_IETF) et ce dernier contient un fichier `messages.json` avec les chaînes de caractères traduites pour la locale correspondante.

Reviewer: ```suggestion
L'internationalisation nécessite de fournir des chaînes traduites pour les différentes locales qu'on souhaite prendre en charge. Pour les extensions, le répertoire `_locales`, présent à la racine de l'extension, contient un sous-répertoire pour chaque locale, nommé grâce à [l'étiquette de langue correspondante](https://fr.wikipedia.org/wiki/%C3%89tiquette_d%27identification_de_langues_IETF) et ce dernier contient un fichier `messages.json` avec les chaînes de caractères traduites pour la locale correspondante.
```

}
```

Ce fichier est du JSON standard, chaque propriété est un objet avec un nom, qui contient une propriété `message` et une propriété `description`. Tous ces éléments sont des chaînes de caractères et `$URL$` est un emplacement de substitution qui sera remplacé par une sous-chaîne lorsque la propriété `notificationContent` sera manipulée par l'extension. Nous verrons comment faire dans la section [Récupération des messages localisés dans le code JavaScript](#récupérer_des_messages_localisés_dans_le_code_javascript).

> **Note :** Pour plus d'informations sur le contenu des fichiers `messages.json`, voir [la référence des formats de messages localisés](/fr/docs/Mozilla/Add-ons/WebExtensions/API/i18n/Locale-Specific_Message_reference).

## Internationalisation du manifeste

Plusieurs actions sont nécessaires pour internationaliser le manifeste (`manifest.json`) de l'extension.

### Récupérer les chaînes localisées dans le manifeste

> [Le manifeste de l'extension](ttps://github.com/mdn/webextensions-examples/blob/main/notify-link-clicks-i18n/manifest.json) contient des chaînes de caractères qui seront visibles de l'utilisatrice et de l'utilisateur, comme le nom et la description de l'extension. Si vous localisez ces chaînes en plaçant les traductions appropriées dans les fichiers `messages.json`, la traduction correspondante sera affichée en fonction des paramètres de locale du navigateur.

Reviewer: ```suggestion
[Le manifeste de l'extension](https://github.com/mdn/webextensions-examples/blob/main/notify-link-clicks-i18n/manifest.json) contient des chaînes de caractères qui seront visibles de l'utilisatrice et de l'utilisateur, comme le nom et la description de l'extension. Si vous internationalisez ces propriétés et placez les traductions appropriées dans les fichiers `messages.json`, la traduction appropriée sera affichée en fonction des paramètres de locale du navigateur.
```

> Pour localiser ces métadonnées, renseignez le manifeste ainsi&nbsp;:

Reviewer: ```suggestion
Pour internationaliser ces propriétés, renseignez le manifeste ainsi&nbsp;:
```

L'internalitionalisation est le processus de généralisation du code, la localisation est le processus de spécialisation. Je pense que ça a plus de sens de garder "internationaliser" ici, et dans un autre commentaire plus haut.

```json
"name": "__MSG_extensionName__",
"description": "__MSG_extensionDescription__",
```

> Ces valeurs spéciales indiqueront au navigateur qu'il convient de rechercher la chaîne localisée correspondante, plutôt que d'utiliser une valeur statique qui serait identique pour chaque locale.

Reviewer: ```suggestion
Ces valeurs spéciales indiqueront au navigateur de rechercher la chaîne localisée correspondante, plutôt que d'utiliser la mème valeur statique pour toutes les locales.
```

> Pour utiliser une référence à un message localisé, composez la chaîne de caractères comme ceci&nbsp;:

Reviewer: ```suggestion
Pour faire référence à un message localisé, composez la chaîne de caractères en concaténant&nbsp;:
```

1. Deux tirets bas (`__`), suivis de
2. La chaîne `MSG`, suivi de
3. Un tiret bas (`_`), suivi du
4. Nom du message visé, tel qu'il est défini dans `messages.json`, suivi de

> 5. Deux tirets bas (`__`)

Reviewer: ```suggestion
1. Deux tirets bas (`__`)
2. La chaîne `MSG`
3. Un tiret bas (`_`)
4. Le nom du message visé, tel qu'il est défini dans `messages.json`
5. Deux tirets bas (`__`)
```

```plain
__MSG_ + messageName + __
```

Par exemple, si on souhaite utiliser le message localisé identifié par la clé `toto`, on écrira ainsi la propriété dans le manifeste&nbsp;:

```plain
"champ_de_manifeste": "__MSG_toto__"

> ```

Reviewer: Peut-être pas nécessaire de rajouter des explications sur ce point qui se comprend facilement.
À mon avis, on pourrait faire l'inverse en supprimant la liste à points et en disant simplement :

La règle générale est `"__MSG_" + nom + "__"`, où `nom` est la chaine utilisée dans les fichiers messages.json.

### Indiquer une locale par défaut

> Dans le manifeste, un autre champ doit être renseigné. Il s'agit de [`default_locale`](/fr/docs/Mozilla/Add-ons/WebExtensions/manifest.json/default_locale), qui indique la locale à utiliser par défaut.

Reviewer: ```suggestion
Dans le manifeste, il est conseillé de renseigner la propriété [`default_locale`](/fr/docs/Mozilla/Add-ons/WebExtensions/manifest.json/default_locale).
```

`should` est simplement un conseil, `doit` est trop fort.

```json
"default_locale": "en"
```

> Cette locale sera celle utilisée si l'extension ne contient pas de chaîne localisée pour la locale courante du navigateur. Si un message n'est pas disponible dans la locale courante du navigateur, ce dernier utilisera la locale par défaut afin de fournir un texte alternatif. Nous verrons ci-après [certaines informations supplémentaires importantes sur la façon dont le navigateur sélectionne les chaînes de caractères effectivement utilisées](#algorithme_de_sélection_de_la_chaîne_de_caractères_localisée).

Reviewer: ```suggestion
Si un message n'est pas disponible dans la locale courante du navigateur, ce dernier cherchera le message dans la locale `default_locale`. Nous verrons ci-après [des informations supplémentaires importantes sur la façon dont le navigateur sélectionne les messages effectivement utilisés](#algorithme_de_sélection_de_la_chaîne_de_caractères_localisée).
```

Le texte en anglais a deux phrases qui disent la même chose, je propose d'en enlever une histoire d'alléger le texte.

## Utiliser du code CSS variant selon la locale

Il est aussi possible d'utiliser des chaînes localisées au sein des fichiers CSS de l'extension. Par exemple, on pourra ainsi construire une règle CSS qui varie selon la locale&nbsp;:

```css
header {
  background-image: url(../images/__MSG_extensionName__/header.png);
}
```

Voir aussi [comment utiliser des messages prédéfinis](#messages_prédéfinis) ci-après qui permet une meilleure gestion dans certains cas.

> ## Récupérer des messages localisés dans le code JavaScript

Reviewer: ```suggestion
## Récupérer des messages localisés en JavaScript
```

> Après avoir mis en place les fichiers contenant les messages et paramétré le manifeste, voyons comment utiliser les messages localisés depuis le code JavaScript afin que l'extension utilise la bonne locale autant que possible. [L'API `i18n`](/fr/docs/Mozilla/Add-ons/WebExtensions/API/i18n) dispose de quatre méthodes&nbsp;:

Reviewer: ```suggestion
Après avoir mis en place les fichiers contenant les messages et paramétré le manifeste, voyons comment utiliser les messages localisés depuis le code JavaScript afin que l'extension utilise la bonne locale autant que possible. [L'API `i18n`](/fr/docs/Mozilla/Add-ons/WebExtensions/API/i18n) dispose de quatre méthodes principales&nbsp;:
```

- [`i18n.getMessage()`](/fr/docs/Mozilla/Add-ons/WebExtensions/API/i18n/getMessage)

>   - : Il s'agit de la méthode que vous utiliserez le plus souvent. Elle permet de récupérer un message traduit à partir de son identifiant. Nous verrons des exemples d'utilisation après.

Reviewer: ```suggestion
  - : Il s'agit de la méthode que vous utiliserez le plus souvent. Elle récupère un message localisé à partir de son identifiant. Nous verrons des exemples d'utilisation après.
```

- [`i18n.getAcceptLanguages()`](/fr/docs/Mozilla/Add-ons/WebExtensions/API/i18n/getAcceptLanguages) et [`i18n.getUILanguage()`](/fr/docs/Mozilla/Add-ons/WebExtensions/API/i18n/getUILanguage)
  - : Ces méthodes peuvent être utilisées pour personnaliser l'interface utilisateur en fonction de la locale (par exemple pour réordonner une liste d'options en mettant les options spécifiques aux langues préférées de l'utilisateur en premier, ou afficher des informations culturelles pertinentes uniquement pour une certaine locale, ou formater les dates affichées en respectant la locale courante du navigateur).
- [`i18n.detectLanguage()`](/fr/docs/Mozilla/Add-ons/WebExtensions/API/i18n/detectLanguage)

>   - : Cette méthode peut être utilisée afin de détecter la langue du contenu saisi par la personne afin de le formater correctement.

Reviewer: ```suggestion
  - : Cette méthode peut être utilisée pour détecter la langue de contenus utilisateurs afin de le formater correctement.
```

"contenu saisi par la personne" implique contenu de l'utilisateur courant, alors qu'on pourrait l'utiliser pour formater du contenu écrit par d'autres utilisateurs.

Dans [l'exemple `notify-link-clicks-i18n`](https://github.com/mdn/webextensions-examples/tree/main/notify-link-clicks-i18n), [le script d'arrière-plan](https://github.com/mdn/webextensions-examples/blob/main/notify-link-clicks-i18n/background-script.js) contient ces lignes&nbsp;:

```js
let title = browser.i18n.getMessage("notificationTitle");
let content = browser.i18n.getMessage("notificationContent", message.url);
```

> La première ligne récupère le message avec l'identifiant `notificationTitle` depuis le fichier `messages.json` le plus pertinent pour la locale courante du navigateur. La seconde ligne est similaire, mais fournit en plus une URL comme deuxième paramètre. C'est cette valeur qui sera utilisée pour remplacer l'emplacement de substitution `$URL$` qui avait été placé dans le champ `message` du champ `notificationContent`&nbsp;:

Reviewer: ```suggestion
La première ligne récupère le message avec l'identifiant `notificationTitle` depuis le fichier `messages.json` le plus pertinent pour la locale courante du navigateur. La seconde ligne est similaire, mais fournit en plus une URL comme deuxième paramètre. Cette valeur sera utilisée pour remplacer l'emplacement de substitution `$URL$` dans le champ `message` du champ `notificationContent`&nbsp;:
```

Par exemple, dans le fichier `en/messages.json`, la chaîne originale du message `notificationContent` est

```
You clicked $URL$.
```

Si le lien sur lequel on a cliqué pointe vers `https://developer.mozilla.org`, après l'appel à [`i18n.getMessage()`](/fr/docs/Mozilla/Add-ons/WebExtensions/API/i18n/getMessage), le contenu du deuxième paramètre sera mis à disposition via l'origine `$1`, et remplacera la sous-chaîne `$URL$` tel que défini dans `"url"`. La chaîne de caractères finalement obtenue sera donc&nbsp;:

```
You clicked https://developer.mozilla.org.
```

### Utilisation directe d'un emplacement de substitution

> Il est possible d'insérer vos variables (`$1`, `$2`, `$3`, etc.) directement dans les chaînes de message. Par exemple nous pourrions réécrire le membre `"notificationContent"` comme ceci&nbsp;:

Reviewer: ```suggestion
Il est possible d'insérer vos variables (`$1`, `$2`, `$3`, etc.) directement dans les chaînes de message. Par exemple nous pourrions réécrire la propriété `"notificationContent"` comme ceci&nbsp;:
```

}
```

> Cela peut sembler plus rapide et moins complexe, mais l'utilisation de `"placeholders"` est considérée comme une meilleure pratique. En effet, disposer du nom de l'emplacement réservé (par exemple `"url"`) et de l'exemple vous aidera à vous souvenir du rôle de cet emplacement. Au contraire, si vous utilisez uniquement `$1`…`$8`, après une semaine, vous aurez plus de difficultés à retrouver les correspondances.

Reviewer: ```suggestion
Cela peut sembler plus rapide et moins complexe, mais l'utilisation de `"placeholders"` est considérée comme une meilleure pratique. En effet, le nom de l'emplacement réservé (par exemple `"url"`) et l'exemple vous aideront à vous souvenir du rôle de cet emplacement. Au contraire, si vous utilisez uniquement `$1`…`$8`, après une semaine, vous aurez plus de difficultés à retrouver les correspondances.
```

}
```

> Dans ce cas, nous plaçons une chaîne statique pour la substitution plutôt que de l'obtenir à partir d'une variable comme `$1`. Cette technique peut s'avérer utile lorsque le fichier contenant les messages est très complexe et qu'il devient nécessaire de séparer différentes valeurs afin de rendre les chaînes plus lisibles dans le fichier. De plus, on pourra accéder à ces valeurs depuis le code JavaScript.

Reviewer: ```suggestion
Dans ce cas, nous plaçons une chaîne statique pour la substitution plutôt que de l'obtenir à partir d'une variable comme `$1`. Cette technique peut s'avérer utile lorsque le fichier de messages est très complexe et qu'il devient nécessaire de séparer différentes valeurs afin de rendre le fichier plus lisible. De plus, ces valeurs sont alors accessibles en JavaScript.
```

En outre, vous pouvez utiliser ces substitutions pour spécifier les parties de la chaîne que vous ne souhaitez pas traduire, telles que les noms de personne ou d'entreprise.

## Algorithme de sélection de la chaîne de caractères localisée

> Les locales peuvent être indiquées à l'aide d'une étiquette de langue simple comme `fr` ou `en`, ou fournies avec un sous-étiquette de région comme `en_US` ou `en_GB`. Lorsqu'on utilise l'API `i18n` afin de récupérer une chaîne de caractères traduite, le moteur du navigateur utilise l'algorithme suivant&nbsp;:

Reviewer: ```suggestion
Les locales peuvent être indiquées à l'aide d'une simple étiquette de langue comme `fr` ou `en`, ou avec en plus une sous-étiquette de région comme `en_US` ou `en_GB`. Lorsqu'on utilise l'API `i18n` afin de récupérer une chaîne de caractères localisée, le navigateur utilise l'algorithme suivant&nbsp;:
```

          ne pouvez pas l'utiliser pour vérifier la propriété <code>id</code> d'un objet
          <a href="/fr/docs/Mozilla/Add-ons/WebExtensions/API/runtime/MessageSender"><code>runtime.MessageSender</code></a>.
        </p>
      </td>
    </tr>
    <tr>
      <td><code>@@ui_locale</code></td>
      <td>La locale courante. Cette chaîne peut être utilisée afin de construire des URL variant en fonction de la locale.
      </td>
    </tr>
    <tr>
      <td><code>@@bidi_dir</code></td>
      <td>
        La direction du texte pour la locale courante. Cette chaîne vaut <code>"ltr"</code> pour les langues écrites de gauche à droite telles que le français ou <code>"rtl"</code> pour les langues écrites de droite à gauche telles que l'arabe.
      </td>
    </tr>
    <tr>
      <td><code>@@bidi_reversed_dir</code></td>
      <td>

>         Si <code>@@bidi_dir</code> vaut <code>"ltr"</code>, alors cette chaîne vaudra <code>"rtl"<code>, sinon elle vaudra <code>"ltr"</code>.

Reviewer: Est-il nécessaire de rajouter les "cette chaine vaudra"? Pour moi, ça alourdi la phrase sans gain.

      </td>
    </tr>
    <tr>
      <td><code>@@bidi_start_edge</code></td>
      <td>
        Si <code>@@bidi_dir</code> vaut <code>"ltr"</code>, alors cette chaîne vaudra <code>"left"<code>, sinon elle vaudra <code>"right"</code>.
      </td>
    </tr>
    <tr>
      <td><code>@@bidi_end_edge</code></td>
      <td>
        Si <code>@@bidi_dir</code> vaut <code>"ltr"</code>, alors cette chaîne vaudra <code>"right"<code>, sinon elle vaudra <code>"left"</code>.
      </td>
    </tr>
  </tbody>
</table>

Ainsi, si nous revenons à l'exemple que nous avions pris pour le code CSS, il serait plus pertinent de l'écrire ainsi&nbsp;:

```css
header {
  background-image: url(../images/__MSG_@@ui_locale__/header.png);
}
```

Grâce à cette écriture, nous pouvons stocker nos images localisées dans des répertoires qui correspondent aux différentes locales prises en charge (`en`, `de`, etc.).

> Prenons un autre exemple où nous utilisons le les messages prédéfinis `@@bidi_*` dans un fichier CSS&nbsp;:

Reviewer: ```suggestion
Prenons un autre exemple où nous utilisons les messages prédéfinis `@@bidi_*` dans un fichier CSS&nbsp;:
```

}
```

> Pour les langues écrites de gauche à droite telles que le français, les déclarations CSS avec les messages prédéfinis ci-dessus seraient extrapolées pour obtenir le code suivant&nbsp;:

Reviewer: ```suggestion
Pour les langues écrites de gauche à droite telles que le français, les déclarations CSS avec les messages prédéfinis ci-dessus se comportent ainsi&nbsp;:
```

## Tester votre extension

> Pour tester la localisation de votre extension, utilisez [Firefox](https://www.mozilla.org/fr/firefox/new/) ou [Firefox Beta](https://www.mozilla.org/fr/firefox/channel/desktop/) qui permettent d'installer des packs de langue.

Reviewer: ```suggestion
Pour tester la localisation de votre extension, utilisez [Firefox](https://www.mozilla.org/fr/firefox/new/) ou [Firefox Beta](https://www.mozilla.org/fr/firefox/channel/desktop/), car ils permettent d'installer des packs de langue.
```

Ensuite, pour chaque locale que vous prenez en charge et que vous voulez tester, suivez les instructions de la page [Utiliser Firefox dans une autre langue](https://support.mozilla.org/fr/kb/utiliser-firefox-dans-autre-langue) afin de changer la locale utilisée pour l'interface utilisateur de Firefox (si vous connaissez les paramètres, allez dans la section Langue et utilisez Choisir des alternatives).

> Lorsque Firefox s'exécute dans la locale à tester, [installez l'extension temporairement](https://extensionworkshop.com/documentation/develop/temporary-installation-in-firefox/). Après avoir installé votre extension, vérifiez dans la page `about:debugging` que votre extension est bien mise en place et que l'icône, le nom et la description apparaissent avec la bonne locale. Vous pouvez également vérifier les métadonnées localisées via la page `about:addons`. Une fois ces vérifications effectuées, utilisez l'extension et ses fonctionnalités pour vérifier que les chaînes localisées sont bien présentes.

Reviewer: ```suggestion
Lorsque Firefox utilise la locale à tester, [installez l'extension temporairement](https://extensionworkshop.com/documentation/develop/temporary-installation-in-firefox/). Après avoir installé votre extension, vérifiez dans la page `about:debugging` que votre extension est bien mise en place et que l'icône, le nom et la description apparaissent avec la bonne locale. Vous pouvez également vérifier les métadonnées localisées via la page `about:addons`. Finalement, utilisez les fonctionnalités de l'extension pour vérifier que les chaînes localisées sont bien présentes.
```
